### PR TITLE
implement open modeler & create new robot

### PR DESCRIPTION
**Pull request by lukashueller**
Cloeses #109 #108 
## Changes
With this PR we add the functionality to create a new robot and open a robot in the modeler. Since the parser for loading the BPMN is not yet finished, only the RobotId is currently displayed on the right in the sidebar. 
We currently have a bug with the searcher but this bug was reported as a note.

## Screenshots
https://user-images.githubusercontent.com/36270527/110770609-cdb7b100-8259-11eb-9b9a-60ed6218ca30.mov

## Checklist before merge

**Developer's responsibilities**
* [ ] **Assign** one or two developers
* [ ] **Change code** if reviewer(s) has/have requested it
* [ ] **Pull request build** has passed
* [ ] **tested locally** (in at least chrome & firefox if frontend)
* [ ] updated the **documentation**
* [ ] added **tests** where necessary

### CODE REVIEW

File: client/src/api/SSOTretrieval.js [hunk 1]
```diff
@@ -7,24 +7,48 @@
...
+  const response = await fetch(requestString);
+  return response;
+};
+
+/**
+ * @description This function renames the robot in SSOT
+ * @param {String} SSOTid - String including the SSotid
+ * @param {String} newName - String with the new RobotName
+ */
+const changeSSOTName = async (SSOTid, newName) => {
```
Comment (WolfgangDaniel) [inline client/src/api/SSOTretrieval.js:21]: Please write SSOTid in camelCase (ssotId) like all the other parameters.

File: client/src/api/SSOTretrieval.js [hunk 2]
```diff
@@ -7,24 +7,48 @@
...
-      const adjustedName = newName.replace(/\s/g, '+');
-      const requestString = `/ssot/renameBot?id=${ SSOTid }&newName=${ adjustedName }`;
-      const response = await fetch(requestString);
-      return response;
-    };
+/**
+ * @description Create a new robot with the specified name for the specified user 
+ * @param {String} newName - String including the user Id
+ */
+const createNewBot = async (userid, newName) => {
```
Comment (WolfgangDaniel) [inline client/src/api/SSOTretrieval.js:42]: userid -> userId (camelCase)

File: client/src/api/SSOTretrieval.js (hunk 1 above)
Comment (SanJSp) [inline client/src/api/SSOTretrieval.js:21]: Please use `ssotId` instead of SSOTid

File: client/src/api/SSOTretrieval.js [hunk 3]
```diff
@@ -7,24 +7,48 @@
...
-  };
+const fetchSSOTsForUser = async (userId) => {
+  const requestString = `/ssot/getAvailableBotsForUser/${userId}`;
+  const response = await fetch(requestString);
+  return response;
+};
+
+/**
+ * @description This function renames the robot in SSOT
+ * @param {String} SSOTid - String including the SSotid
```
Comment (SanJSp) [inline client/src/api/SSOTretrieval.js:18]: Please use only one writing style for the ssot

File: client/src/components/content/BpmnModeler/BpmnModeler.jsx [hunk 4]
```diff
@@ -75,9 +74,13 @@ const BpmnModeler = () => {
...
+      <ModelerSidebar robotId={robotId} modeler={modeler} getRobotFile={downloadRobotFile} />
     </Layout>
   );
 };
 
 export default BpmnModeler;
+
+BpmnModeler.propTypes = {
+  robotId: PropTypes.string.isRequired
+};
```
Comment (WolfgangDaniel) [inline client/src/components/content/BpmnModeler/BpmnModeler.jsx:86]: I am not quite sure but I think export should always be in the last line.

File: client/src/api/SSOTretrieval.js [hunk 5]
```diff
@@ -7,24 +7,48 @@
...
+/**
+ * @description Create a new robot with the specified name for the specified user 
+ * @param {String} newName - String including the user Id
+ */
+const createNewBot = async (userid, newName) => {
+  const adjustedName = newName.replace(/\s/g, '+');
+  const requestString = `/ssot/createNewBot?userid=${userid}&botName=${adjustedName}`;
+  const response = await fetch(requestString);
+  return response;
+};
```
Comment (SanJSp) [inline client/src/api/SSOTretrieval.js:47]: Why is the use of newName necessary here? Shouldn't the name be set with the call that we use to change the name?

File: client/src/components/content/RobotContainer/CreateRobotContainer.jsx [hunk 6]
```diff
@@ -35,3 +36,7 @@ const CreateRobotContainer = () => {
     )
 };
 export default CreateRobotContainer;
+
+CreateRobotContainer.propTypes = {
+    createNewRobot: PropTypes.string.isRequired,
+};
```
Comment (WolfgangDaniel) [inline client/src/components/content/RobotContainer/CreateRobotContainer.jsx:42]: Export should start after propTypes :) I just looked it up on reactjs.org and found this example:
grafik

File: client/src/components/content/BpmnModeler/BpmnModeler.jsx [hunk 7]
```diff
@@ -75,9 +74,13 @@ const BpmnModeler = () => {
           <div className={styles['bpmn-modeler-container']} id='bpmnview' />
         </div>
       </Content>
-      <ModelerSidebar modeler={modeler} getRobotFile={downloadRobotFile} />
+      <ModelerSidebar robotId={robotId} modeler={modeler} getRobotFile={downloadRobotFile} />
```
Comment (SanJSp) [inline client/src/components/content/BpmnModeler/BpmnModeler.jsx:77]: Do we need the displaying of the bot id in the sidebar or was this only for testing purposes?

File: client/src/components/content/RobotContainer/CreateRobotContainer.jsx (hunk 6 above)
Comment (WolfgangDaniel) [reply client/src/components/content/RobotContainer/CreateRobotContainer.jsx:42]: Please change this for all the documents where you used propTypes

File: client/src/components/content/BpmnModeler/BpmnModeler.jsx (hunk 4 above)
Comment (SanJSp) [inline client/src/components/content/BpmnModeler/BpmnModeler.jsx:86]: That looks cool, but I dont understand it! Can you explain it or maybe link me to some resource I can read through?

File: client/src/components/pages/Modeler/Modeler.jsx [hunk 8]
```diff
@@ -3,21 +3,26 @@ import { Layout } from 'antd';
 import BpmnModeler from '../../content/BpmnModeler/BpmnModeler';
 import HeaderNavbar from '../../content/HeaderNavbar/HeaderNavbar';
 
+
```
Comment (SanJSp) [inline client/src/components/pages/Modeler/Modeler.jsx:6]: empty line

File: client/src/components/pages/Modeler/Modeler.jsx [hunk 9]
```diff
@@ -3,21 +3,26 @@ import { Layout } from 'antd';
...
-  <div>
-    <Layout>
-      <HeaderNavbar selectedKey={2} />
-      <BpmnModeler />
-      <Footer>Fußzeile</Footer>
-    </Layout>
-  </div>
-);
+const Modeler = (match) => {
+  const { robotId } = match.match.params;
```
Comment (SanJSp) [inline client/src/components/pages/Modeler/Modeler.jsx:15]: What does this do? :)

File: server/controllers/ssotRetrievalController.js [hunk 10]
```diff
@@ -16,34 +16,34 @@ exports.getSingleSourceOfTruth = async (req, res) => {
     }
 };
 
-// GET /getAvailableBotsForUser/:userid
+// GET /getAvailableBotsForUser/78d09f66d2ed466cf20b06f7
```
Comment (SanJSp) [inline server/controllers/ssotRetrievalController.js:19]: why was this comment to explain the schema changed?

File: server/controllers/ssotRetrievalController.js [hunk 11]
```diff
@@ -55,17 +55,18 @@ exports.getBotList = async (req, res) => {
     }
 };
 
-// GET /renameBot?id=Browser&newName=Open+Browser
+// GET /renameBot?id=78d09f66d2ed466cf20b06f7&newName=Bot+Browser
```
Comment (SanJSp) [inline server/controllers/ssotRetrievalController.js:58]: again, why the change of the id?

File: server/controllers/ssotRetrievalController.js [hunk 12]
```diff
@@ -77,3 +78,79 @@ exports.renameBot = async (req, res) => {
...
+        const returnObj = {
+            robotName: nameWithEmptyspace,
+            robotId: ssot.id
+        };
+
+        res.send(returnObj);
+    } catch (err) {
+        console.error(err);
+    }
+};
```
Comment (SanJSp) [inline server/controllers/ssotRetrievalController.js:156]: newline @ EOF

File: client/src/components/pages/RobotOverview/RobotOverview.jsx [hunk 13]
```diff
@@ -60,6 +61,24 @@ const RobotOverview = () => {
...
+   */
+  const createNewRobot = () => {
+    const robotName = 'New Robot';
+    createNewBot(userId, robotName)
+      .then((response) => response.json())
+      .then((newRobot) => {
+        const newRobotList = [...robotList]
+        newRobotList.unshift(newRobot);
+        setRobotList([]);
+        setRobotList(newRobotList);
```
Comment (WolfgangDaniel) [inline client/src/components/pages/RobotOverview/RobotOverview.jsx:75]: We don't need setRobotList([]) because setRobotList(newRobotList) just _replaces_ the current value of the state variable

File: server/routes/ssot.js [hunk 14]
```diff
@@ -7,10 +7,12 @@ const parsingController = require('../controllers/ssotParsingController');
...
 
 /// SSOTPARSING ROUTES ///
-
 router.get('/parser/get-robot-code', parsingController.getRobotCode);
 router.get('/get/:id', retrievalController.getSingleSourceOfTruth);
 router.get('/getAvailableBotsForUser/:userid', retrievalController.getBotList);
 router.get('/renameBot', retrievalController.renameBot);
+router.get('/retrieveMetadataForBot/:botId', retrievalController.retrieveBotMetadata);
+router.get('/shareBotWithUser', retrievalController.shareBotWithUser);
+router.get('/createNewBot', retrievalController.createNewBot);
```
Comment (SanJSp) [inline server/routes/ssot.js:16]: Can we maybe think of new names here? Something like /bot/new and /bot/edit/rename would be good for categorizing the routes

**Review by SanJSp — CHANGES_REQUESTED** [review]

File: server/controllers/ssotRetrievalController.js [hunk 15]
```diff
@@ -16,34 +16,34 @@ exports.getSingleSourceOfTruth = async (req, res) => {
...
+            { _id: { $in: SSOTIds } },
             {
                 robotMetadata: 1
             }
         ).exec();
-        
+
         const SSOTs = [];
         availableSSOTs.forEach((singleSSOT) => {
             SSOTs.push(singleSSOT.robotMetadata);
```
Comment (WolfgangDaniel) [inline server/controllers/ssotRetrievalController.js:49]: I would rather use camelCase consistently and change SSOTIds to ssotIds

File: server/controllers/ssotRetrievalController.js [hunk 16]
```diff
@@ -77,3 +78,79 @@ exports.renameBot = async (req, res) => {
...
+            robotId: ssot.id,
+            userId: usableUserId
+        });
+
+        const returnObj = {
+            robotName: nameWithEmptyspace,
+            robotId: ssot.id
+        };
+
+        res.send(returnObj);
```
Comment (WolfgangDaniel) [inline server/controllers/ssotRetrievalController.js:152]: You used bot instead of robot. Does this have a specific reason? Because so far I think we were always referring to them as robot. I don't mind using bot but I think it should be consistent

**Review by WolfgangDaniel — CHANGES_REQUESTED** [review]
Great job Lukas! Really like the implementation. I have just requested some minor changes, mostly regarding naming.

File: client/src/components/content/BpmnModeler/BpmnModeler.jsx (hunk 4 above)
Comment (WolfgangDaniel) [reply client/src/components/content/BpmnModeler/BpmnModeler.jsx:86]: https://reactjs.org/docs/typechecking-with-proptypes.html

File: client/src/components/content/BpmnModeler/BpmnModeler.jsx (hunk 7 above)
Comment (WolfgangDaniel) [reply client/src/components/content/BpmnModeler/BpmnModeler.jsx:77]: I think this is a good way to show that it works as long as the parsers aren't implemented and we can't look at the model to see if it works

Comment (WolfgangDaniel) [reply client/src/components/content/BpmnModeler/BpmnModeler.jsx:77]: Longterm I would still put the robot name at least in the sidebar or somewhere else on the modeler page :)

File: server/routes/ssot.js (hunk 14 above)
Comment (WolfgangDaniel) [reply server/routes/ssot.js:16]: As mentioned in one of my comments below I think we should also use 'robot' or 'bot' consistently

File: client/src/api/SSOTretrieval.js (hunk 1 above)
Comment (lukashueller) [reply client/src/api/SSOTretrieval.js:21]: ✅ I've renamed all occurrences of `SSOT | Ssot | SSoT | ssot`  to `Ssot or ssot` (depending on context)

File: client/src/api/SSOTretrieval.js (hunk 3 above)
Comment (lukashueller) [reply client/src/api/SSOTretrieval.js:18]: same fix as above

File: client/src/api/SSOTretrieval.js (hunk 5 above)
Comment (lukashueller) [reply client/src/api/SSOTretrieval.js:47]: In principle you are right, however we have directly prepared the backend and also this frontend function to receive a parameter from us. This makes it easier to implement the background functionality behind the CreateNewRobotPopup. (#169)

File: client/src/components/content/BpmnModeler/BpmnModeler.jsx (hunk 7 above)
Comment (lukashueller) [reply client/src/components/content/BpmnModeler/BpmnModeler.jsx:77]: I agree with Daniel 100%. The `RobotId` is just to show that the right bot is taken and the `RobotId` is ready to use. In the long run, the robot name should be displayed there at the top (or at another place on the page) (unfortunately, I could not implement this as quickly as planned).

File: client/src/components/content/BpmnModeler/BpmnModeler.jsx (hunk 4 above)
Comment (lukashueller) [reply client/src/components/content/BpmnModeler/BpmnModeler.jsx:86]: Thanks Daniel! (@SanJSp I've implemented it in every frontend-function where necessary to satisfy the linter)

File: client/src/components/pages/Modeler/Modeler.jsx (hunk 8 above)
Comment (lukashueller) [reply client/src/components/pages/Modeler/Modeler.jsx:6]: removed ✅

File: client/src/components/pages/Modeler/Modeler.jsx (hunk 9 above)
Comment (lukashueller) [reply client/src/components/pages/Modeler/Modeler.jsx:15]: This piece of code retrieves the passed RobotId (which was passed via the routing of the React router).

`RobotContainer.jsx` (the edit-Button links to the modeler)
image

the final `route` is shown in the `URL`
image

`App.js` (responsible for all routes)
image

If you want more information about routing with params, have a look at [this article](https://betterprogramming.pub/how-to-pass-multiple-route-parameters-in-a-react-url-path-4b919de0abbe) [section "Routing Parameters"]. @SanJSp @WolfgangDaniel

File: server/controllers/ssotRetrievalController.js (hunk 10 above)
Comment (lukashueller) [reply server/controllers/ssotRetrievalController.js:19]: Erik wanted to put example calls in the comments, so he replaced the variable name with a real userId.

File: server/controllers/ssotRetrievalController.js (hunk 11 above)
Comment (lukashueller) [reply server/controllers/ssotRetrievalController.js:58]: same as above

File: server/controllers/ssotRetrievalController.js (hunk 12 above)
Comment (lukashueller) [reply server/controllers/ssotRetrievalController.js:156]: fixed

File: server/routes/ssot.js (hunk 14 above)
Comment (lukashueller) [reply server/routes/ssot.js:16]: This is part of issues #155 , to build such a schema that allows uniform and easier accesses to our API / backend.

File: client/src/api/SSOTretrieval.js (hunk 1 above)
Comment (lukashueller) [reply client/src/api/SSOTretrieval.js:21]: done ✅

File: client/src/api/SSOTretrieval.js (hunk 2 above)
Comment (lukashueller) [reply client/src/api/SSOTretrieval.js:42]: done ✅

File: client/src/components/content/BpmnModeler/BpmnModeler.jsx (hunk 4 above)
Comment (lukashueller) [reply client/src/components/content/BpmnModeler/BpmnModeler.jsx:86]: Good point. ESLINT says nothing about this. Nevertheless, I have changed all occurrences in the code and also added it to our coding standards.

File: client/src/components/content/RobotContainer/CreateRobotContainer.jsx (hunk 6 above)
Comment (lukashueller) [reply client/src/components/content/RobotContainer/CreateRobotContainer.jsx:42]: Ahh :D just read now. Awesome research, thank you! Had it already changed anyway.

File: server/controllers/ssotRetrievalController.js (hunk 15 above)
Comment (lukashueller) [reply server/controllers/ssotRetrievalController.js:49]: fixed

File: server/controllers/ssotRetrievalController.js (hunk 16 above)
Comment (lukashueller) [reply server/controllers/ssotRetrievalController.js:152]: Fair Point. Hadn't noticed that at all, those code posts came entirely from Erik. At this point also the request to you @kej-jay to use robot instead of bot in the future.
Anyway: Fixed

File: server/controllers/ssotRetrievalController.js (hunk 10 above)
Comment (kej-jay) [reply server/controllers/ssotRetrievalController.js]: We usually had real examples in those comments, so I wanted to keep that style

File: server/controllers/ssotRetrievalController.js (hunk 11 above)
Comment (kej-jay) [reply server/controllers/ssotRetrievalController.js:58]: see above

File: client/src/components/pages/RobotOverview/RobotOverview.jsx (hunk 13 above)
Comment (kej-jay) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:75]: @lukashueller

Comment (lukashueller) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:75]: unfortunately, this is not the right case. When I remove this workaround, React only re-renders the last box in the GUI (adds a box at the back). Therefore, without this line of code, I don't get the desired result that the new robots are displayed first on the page. (As @SanJSp had requested)

Do you maybe have another idea how to solve the problem? Otherwise I would leave the workaround in there for now.
@WolfgangDaniel

File: client/src/components/pages/Modeler/Modeler.jsx (hunk 9 above)
Comment (SanJSp) [reply client/src/components/pages/Modeler/Modeler.jsx:14]: pretty cool :) thanks!

**Review by SanJSp — APPROVED** [review]

File: client/src/components/pages/RobotOverview/RobotOverview.jsx (hunk 13 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:75]: Interesting. I just looked it up and it seems like unshift might cause the problem. I would try writing the 4 lines as one line like this: setRobotList( robotList => [newRobot, ...robotList])
That should work. If you are intersted, here is a link to a medium article about that topic [](https://javascript.plainenglish.io/how-to-add-to-an-array-in-react-state-3d08ddb2e1dc)

**Review by WolfgangDaniel — APPROVED** [review]
You can merge imo but please at first try to implement my suggestion regarding the setRobotList([]) call in RobotOverview

File: client/src/components/pages/RobotOverview/RobotOverview.jsx (hunk 13 above)
Comment (lukashueller) [reply client/src/components/pages/RobotOverview/RobotOverview.jsx:75]: I was not able to fix it. Maybe lets try it in a bigger refactoring.
